File: api/services/facebook_poster.py
import re
import urllib.parse
import requests
from api.config import Config

class FacebookPoster:
    def __init__(self):
        # Default FB Graph API version from config, e.g. "v19.0"
        self.base_url = f"https://graph.facebook.com/{Config.FB_GRAPH_API_VERSION}"

    def _request(self, method: str, path: str, params: dict) -> dict:
        url = f"{self.base_url}{path}"
        try:
            if method == 'GET':
                res = requests.get(url, params=params, timeout=30)
            else:
                res = requests.post(url, data=params, timeout=30)
            return {'httpCode': res.status_code, 'data': res.json()}
        except Exception as e:
            return {'httpCode': 500, 'data': {'error': {'message': str(e)}}}

    def get_page_followers(self, page_id: str, token: str) -> dict:
        result = self._request('GET', f"/{page_id}", {
            'fields': 'followers_count,fan_count,name',
            'access_token': token
        })

        if result['httpCode'] != 200 or 'id' not in result['data']:
            msg = result['data'].get('error', {}).get('message') or f"HTTP {result['httpCode']}"
            return {'success': False, 'message': msg}

        followers = int(result['data'].get('followers_count') or result['data'].get('fan_count') or 0)
        return {
            'success': True,
            'followers': followers,
            'name': result['data'].get('name')
        }

    def get_instagram_followers(self, page_id: str, token: str, cached_ig_id: str = None) -> dict:
        ig_id = cached_ig_id
        if not ig_id:
            link_result = self._request('GET', f"/{page_id}", {
                'fields': 'instagram_business_account',
                'access_token': token
            })
            ig_id = link_result.get('data', {}).get('instagram_business_account', {}).get('id')
            if not ig_id:
                return {'success': False, 'message': 'No Instagram Business Account Linked To This Facebook Page.'}

        result = self._request('GET', f"/{ig_id}", {
            'fields': 'followers_count,username',
            'access_token': token
        })

        if result['httpCode'] != 200 or 'id' not in result['data']:
            msg = result['data'].get('error', {}).get('message') or f"HTTP {result['httpCode']}"
            return {'success': False, 'message': msg}

        return {
            'success': True,
            'followers': int(result['data'].get('followers_count') or 0),
            'ig_business_account_id': ig_id,
            'username': result['data'].get('username')
        }

    def get_recent_posts(self, page_id: str, token: str, limit: int = 10) -> dict:
        result = self._request('GET', f"/{page_id}/posts", {
            'fields': 'id,message,full_picture,created_time',
            'limit': limit,
            'access_token': token
        })

        if result['httpCode'] != 200 or 'data' not in result['data']:
            msg = result['data'].get('error', {}).get('message') or f"HTTP {result['httpCode']}"
            return {'success': False, 'message': msg}

        return {'success': True, 'posts': result['data']['data']}

    def get_post(self, post_id: str, token: str) -> dict:
        result = self._request('GET', f"/{post_id}", {
            'fields': 'message,full_picture,created_time',
            'access_token': token
        })

        if result['httpCode'] != 200 or 'error' in result['data']:
            msg = result['data'].get('error', {}).get('message') or f"HTTP {result['httpCode']}"
            return {'success': False, 'message': msg}

        return {
            'success': True,
            'message': result['data'].get('message', ''),
            'image_url': result['data'].get('full_picture')
        }

    def publish_to_page(self, page_id: str, token: str, message: str, image_url: str = None, video_url: str = None) -> dict:
        if video_url:
            result = self._request('POST', f"/{page_id}/videos", {
                'file_url': video_url,
                'description': message,
                'access_token': token
            })
            post_id_field = 'id'
        elif image_url:
            result = self._request('POST', f"/{page_id}/photos", {
                'url': image_url,
                'caption': message,
                'access_token': token
            })
            post_id_field = 'post_id'
        else:
            result = self._request('POST', f"/{page_id}/feed", {
                'message': message,
                'access_token': token
            })
            post_id_field = 'id'

        if result['httpCode'] != 200 or not result['data'].get(post_id_field):
            msg = result['data'].get('error', {}).get('message') or f"HTTP {result['httpCode']}"
            return {'success': False, 'message': msg}

        return {'success': True, 'fb_post_id': result['data'][post_id_field]}

    def exchange_for_long_lived_token(self, short_lived_token: str) -> dict:
        result = self._request('GET', '/oauth/access_token', {
            'grant_type': 'fb_exchange_token',
            'client_id': Config.FB_APP_ID,
            'client_secret': Config.FB_APP_SECRET,
            'fb_exchange_token': short_lived_token
        })

        if result['httpCode'] != 200 or not result['data'].get('access_token'):
            msg = result['data'].get('error', {}).get('message') or f"HTTP {result['httpCode']}"
            return {'success': False, 'message': msg}

        return {'success': True, 'long_lived_token': result['data']['access_token']}

    def get_page_access_token(self, page_id: str, long_lived_user_token: str) -> dict:
        result = self._request('GET', f"/{page_id}", {
            'fields': 'access_token,name',
            'access_token': long_lived_user_token
        })

        if result['httpCode'] != 200 or not result['data'].get('access_token'):
            msg = result['data'].get('error', {}).get('message') or f"HTTP {result['httpCode']}"
            return {'success': False, 'message': msg}

        return {
            'success': True,
            'page_access_token': result['data']['access_token'],
            'name': result['data'].get('name')
        }

    def extract_post_id(self, url: str) -> str:
        url = url.strip()

        # https://www.facebook.com/{page}/posts/{post_id}
        m = re.search(r'/posts/([a-zA-Z0-9]+)', url)
        if m:
            return m.group(1)

        # https://www.facebook.com/{page}/videos/{video_id}
        m = re.search(r'/videos/(\d+)', url)
        if m:
            return m.group(1)

        # https://www.facebook.com/reel/{reel_id}
        m = re.search(r'/reel/(\d+)', url)
        if m:
            return m.group(1)

        # story / permalink
        parsed = urllib.parse.urlparse(url)
        params = urllib.parse.parse_qs(parsed.query)
        if 'story_fbid' in params and 'id' in params:
            return f"{params['id'][0]}_{params['story_fbid'][0]}"

        return None

File: includes/FacebookPoster.php
<?php
require_once __DIR__ . '/../config.php';

class FacebookPoster
{
    private string $baseUrl;

    public function __construct()
    {
        $this->baseUrl = "https://graph.facebook.com/" . FB_GRAPH_API_VERSION;
    }

    private function request(string $method, string $path, array $params): array
    {
        $url = "{$this->baseUrl}{$path}";
        $ch = curl_init();

        if ($method === 'GET') {
            curl_setopt($ch, CURLOPT_URL, $url . '?' . http_build_query($params));
        } else {
            curl_setopt($ch, CURLOPT_URL, $url);
            curl_setopt($ch, CURLOPT_POST, true);
            curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($params));
        }

        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, 30);
        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        $data = json_decode($response, true);
        return ['httpCode' => $httpCode, 'data' => $data];
    }

    /**
     * Official Graph API follower count — used for dealerships that have granted
     * real Page admin access (see dealerships.fb_page_access_token), instead of
     * scraping. No quota/blocking concerns like Bright Data/RapidAPI/ScrapeCreators.
     */
    public function getPageFollowers(string $pageId, string $token): array
    {
        $result = $this->request('GET', "/{$pageId}", [
            'fields' => 'followers_count,fan_count,name',
            'access_token' => $token,
        ]);

        if ($result['httpCode'] !== 200 || !isset($result['data']['id'])) {
            $message = $result['data']['error']['message'] ?? ('HTTP ' . $result['httpCode']);
            return ['success' => false, 'message' => $message];
        }

        return [
            'success' => true,
            'followers' => (int)($result['data']['followers_count'] ?? $result['data']['fan_count'] ?? 0),
            'name' => $result['data']['name'] ?? null,
        ];
    }

    /**
     * Instagram data goes through the linked Facebook Page's own token — no
     * separate Instagram login/token needed, as long as the IG account is a
     * Business/Creator account connected to this Page. $cachedIgId (dealerships.
     * ig_business_account_id) skips the resolve step once it's known.
     */
    public function getInstagramFollowers(string $pageId, string $token, ?string $cachedIgId = null): array
    {
        $igId = $cachedIgId;
        if (!$igId) {
            $linkResult = $this->request('GET', "/{$pageId}", [
                'fields' => 'instagram_business_account',
                'access_token' => $token,
            ]);
            $igId = $linkResult['data']['instagram_business_account']['id'] ?? null;
            if (!$igId) {
                return ['success' => false, 'message' => 'No Instagram Business Account Linked To This Facebook Page.'];
            }
        }

        $result = $this->request('GET', "/{$igId}", [
            'fields' => 'followers_count,username',
            'access_token' => $token,
        ]);

        if ($result['httpCode'] !== 200 || !isset($result['data']['id'])) {
            $message = $result['data']['error']['message'] ?? ('HTTP ' . $result['httpCode']);
            return ['success' => false, 'message' => $message];
        }

        return [
            'success' => true,
            'followers' => (int)($result['data']['followers_count'] ?? 0),
            'ig_business_account_id' => $igId,
            'username' => $result['data']['username'] ?? null,
        ];
    }

    public function getRecentPosts(string $pageId, string $token, int $limit = 10): array
    {
        $result = $this->request('GET', "/{$pageId}/posts", [
            'fields' => 'id,message,full_picture,created_time',
            'limit' => $limit,
            'access_token' => $token,
        ]);

        if ($result['httpCode'] !== 200 || !isset($result['data']['data'])) {
            $message = $result['data']['error']['message'] ?? ('HTTP ' . $result['httpCode']);
            return ['success' => false, 'message' => $message];
        }

        return ['success' => true, 'posts' => $result['data']['data']];
    }

    public function getPost(string $postId, string $token): array
    {
        $result = $this->request('GET', "/{$postId}", [
            'fields' => 'message,full_picture,created_time',
            'access_token' => $token,
        ]);

        if ($result['httpCode'] !== 200 || isset($result['data']['error'])) {
            $errorMessage = $result['data']['error']['message'] ?? ('HTTP ' . $result['httpCode']);
            return ['success' => false, 'message' => $errorMessage];
        }

        return [
            'success' => true,
            'message' => $result['data']['message'] ?? '',
            'image_url' => $result['data']['full_picture'] ?? null,
        ];
    }

    /**
     * Posting with 'link' set to the source post's own facebook.com permalink
     * (to get Facebook's embedded "shared post" card look) was tried and
     * reverted — Facebook rejects it with "must be granted before impersonating
     * a user's page" unless the Page Access Token carries extra permissions
     * (pages_read_engagement/pages_manage_metadata/etc.) that this token
     * doesn't have. Reshare-look posting goes through Zapier instead (see
     * send_to_zapier.php); this always posts as fresh content.
     */
    public function publishToPage(string $pageId, string $token, string $message, ?string $imageUrl, ?string $videoUrl = null): array
    {
        if (!empty($videoUrl)) {
            $result = $this->request('POST', "/{$pageId}/videos", [
                'file_url' => $videoUrl,
                'description' => $message,
                'access_token' => $token,
            ]);
            $postIdField = 'id';
        } elseif (!empty($imageUrl)) {
            $result = $this->request('POST', "/{$pageId}/photos", [
                'url' => $imageUrl,
                'caption' => $message,
                'access_token' => $token,
            ]);
            $postIdField = 'post_id';
        } else {
            $result = $this->request('POST', "/{$pageId}/feed", [
                'message' => $message,
                'access_token' => $token,
            ]);
            $postIdField = 'id';
        }

        if ($result['httpCode'] !== 200 || empty($result['data'][$postIdField])) {
            $errorMessage = $result['data']['error']['message'] ?? ('HTTP ' . $result['httpCode']);
            return ['success' => false, 'message' => $errorMessage];
        }

        return ['success' => true, 'fb_post_id' => $result['data'][$postIdField]];
    }

    /**
     * Exchanges a short-lived token (the one Graph API Explorer / a dealership's
     * login hands out, ~1-2hr life) for a long-lived one (~60 days). Needs
     * FB_APP_ID/FB_APP_SECRET since only the app that issued the original token
     * can request this exchange.
     */
    public function exchangeForLongLivedToken(string $shortLivedToken): array
    {
        $result = $this->request('GET', '/oauth/access_token', [
            'grant_type' => 'fb_exchange_token',
            'client_id' => FB_APP_ID,
            'client_secret' => FB_APP_SECRET,
            'fb_exchange_token' => $shortLivedToken,
        ]);

        if ($result['httpCode'] !== 200 || empty($result['data']['access_token'])) {
            $message = $result['data']['error']['message'] ?? ('HTTP ' . $result['httpCode']);
            return ['success' => false, 'message' => $message];
        }

        return ['success' => true, 'long_lived_token' => $result['data']['access_token']];
    }

    /**
     * A Page token derived from a long-lived user token doesn't carry its own
     * expiry — it stays valid until the page role, app connection, or user
     * password changes. This is the "permanent" token target_pages.php expects.
     */
    public function getPageAccessToken(string $pageId, string $longLivedUserToken): array
    {
        $result = $this->request('GET', "/{$pageId}", [
            'fields' => 'access_token,name',
            'access_token' => $longLivedUserToken,
        ]);

        if ($result['httpCode'] !== 200 || empty($result['data']['access_token'])) {
            $message = $result['data']['error']['message'] ?? ('HTTP ' . $result['httpCode']);
            return ['success' => false, 'message' => $message];
        }

        return [
            'success' => true,
            'page_access_token' => $result['data']['access_token'],
            'name' => $result['data']['name'] ?? null,
        ];
    }

    /**
     * Parses common Facebook post URL shapes into a Graph API-usable ID.
     * Returns null if the URL doesn't match a known pattern.
     */
    public function extractPostId(string $url): ?string
    {
        $url = trim($url);

        // https://www.facebook.com/{page}/posts/{post_id}
        if (preg_match('#/posts/([a-zA-Z0-9]+)#', $url, $m)) {
            return $m[1];
        }

        // https://www.facebook.com/{page}/videos/{video_id}
        if (preg_match('#/videos/(\d+)#', $url, $m)) {
            return $m[1];
        }

        // https://www.facebook.com/reel/{reel_id}
        if (preg_match('#/reel/(\d+)#', $url, $m)) {
            return $m[1];
        }

        // https://www.facebook.com/story.php?story_fbid=X&id=Y  or  /permalink.php?story_fbid=X&id=Y
        $query = parse_url($url, PHP_URL_QUERY);
        if ($query) {
            parse_str($query, $params);
            if (!empty($params['story_fbid']) && !empty($params['id'])) {
                return $params['id'] . '_' . $params['story_fbid'];
            }
        }

        return null;
    }
}
